File: src/main/java/net/flyingfishflash/loremlist/persistence/LrmItemRepositoryRdbms.java
package net.flyingfishflash.loremlist.persistence;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import kotlinx.datetime.Clock;
import kotlinx.datetime.ConvertersKt;
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItem;
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemRepository;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

@Repository
class LrmItemRepositoryRdbms implements LrmItemRepository {

  private static final String ITEM_COLUMNS =
      "i.id AS item_id, i.name AS item_name, i.description AS item_description, "
          + "i.owner AS item_owner, i.created AS item_created, i.creator AS item_creator, "
          + "i.updated AS item_updated, i.updater AS item_updater";

  private static final String FIND_ITEMS_SQL =
      "SELECT "
          + ITEM_COLUMNS
          + ", l.id AS list_id, l.name AS list_name "
          + "FROM item i LEFT JOIN list_item li ON li.item_id = i.id LEFT JOIN list l ON l.id = li.list_id";

  private static final RowMapper<ItemRow> ITEM_ROW_MAPPER =
      (rs, rowNum) ->
          new ItemRow(
              (UUID) rs.getObject("item_id"),
              rs.getString("item_name"),
              rs.getString("item_description"),
              rs.getString("item_owner"),
              rs.getObject("item_created", Instant.class),
              rs.getString("item_creator"),
              rs.getObject("item_updated", Instant.class),
              rs.getString("item_updater"),
              (UUID) rs.getObject("list_id"),
              rs.getString("list_name"));

  private final JdbcClient jdbcClient;

  LrmItemRepositoryRdbms(JdbcClient jdbcClient) {
    this.jdbcClient = jdbcClient;
  }

  @Override
  public long countByOwner(String owner) {
    return jdbcClient
        .sql("SELECT COUNT(*) FROM item WHERE owner = :owner")
        .param("owner", owner)
        .query(Long.class)
        .single();
  }

  @Override
  public int delete() {
    return jdbcClient.sql("DELETE FROM item").update();
  }

  @Override
  public int deleteById(Set<UUID> ids) {
    return jdbcClient.sql("DELETE FROM item WHERE id IN (:ids)").param("ids", ids).update();
  }

  @Override
  public int deleteByOwnerAndId(UUID id, String owner) {
    return jdbcClient
        .sql("DELETE FROM item WHERE id = :id AND owner = :owner")
        .param("id", id)
        .param("owner", owner)
        .update();
  }

  @Override
  public List<LrmItem> findByOwner(String owner) {
    List<ItemRow> rows =
        jdbcClient
            .sql(FIND_ITEMS_SQL + " WHERE i.owner = :owner")
            .param("owner", owner)
            .query(ITEM_ROW_MAPPER)
            .list();
    return mapRowsToLrmItems(rows);
  }

  @Override
  public LrmItem findByOwnerAndIdOrNull(UUID id, String owner) {
    List<ItemRow> rows =
        jdbcClient
            .sql(FIND_ITEMS_SQL + " WHERE i.owner = :owner AND i.id = :id")
            .param("owner", owner)
            .param("id", id)
            .query(ITEM_ROW_MAPPER)
            .list();
    List<LrmItem> lrmItems = mapRowsToLrmItems(rows);
    if (lrmItems.size() > 1) {
      throw new IllegalStateException("This query should return no more than one record.");
    }
    return lrmItems.isEmpty() ? null : lrmItems.get(0);
  }

  @Override
  public List<LrmItem> findByOwnerAndHavingNoListAssociations(String owner) {
    return jdbcClient
        .sql(
            "SELECT "
                + ITEM_COLUMNS
                + ", CAST(NULL AS UUID) AS list_id, CAST(NULL AS VARCHAR) AS list_name "
                + "FROM item i LEFT JOIN list_item li ON li.item_id = i.id "
                + "WHERE i.owner = :owner AND li.item_id IS NULL")
        .param("owner", owner)
        .query(ITEM_ROW_MAPPER)
        .list()
        .stream()
        .map(LrmItemRepositoryRdbms::toLrmItem)
        .toList();
  }

  @Override
  public List<LrmItem> findByOwnerAndHavingNoListAssociations(String owner, UUID listId) {
    return jdbcClient
        .sql(
            "SELECT "
                + ITEM_COLUMNS
                + ", CAST(NULL AS UUID) AS list_id, CAST(NULL AS VARCHAR) AS list_name "
                + "FROM item i LEFT JOIN list_item li ON li.item_id = i.id "
                + "WHERE (i.owner = :owner AND li.list_id <> :listId) OR li.list_id IS NULL")
        .param("owner", owner)
        .param("listId", listId)
        .query(ITEM_ROW_MAPPER)
        .list()
        .stream()
        .map(LrmItemRepositoryRdbms::toLrmItem)
        .toList();
  }

  @Override
  public List<UUID> findIdsByOwnerAndIds(List<UUID> itemIdCollection, String owner) {
    return jdbcClient
        .sql("SELECT id FROM item WHERE id IN (:ids) AND owner = :owner")
        .param("ids", itemIdCollection)
        .param("owner", owner)
        .query((rs, rowNum) -> (UUID) rs.getObject("id"))
        .list();
  }

  @Override
  public Set<UUID> notFoundByOwnerAndId(List<UUID> itemIdCollection, String owner) {
    Set<UUID> foundIds = new LinkedHashSet<>(findIdsByOwnerAndIds(itemIdCollection, owner));
    Set<UUID> notFound = new LinkedHashSet<>(itemIdCollection);
    notFound.removeAll(foundIds);
    return notFound;
  }

  @Override
  public UUID insert(LrmItem lrmItem) {
    jdbcClient
        .sql(
            "INSERT INTO item (id, name, description, owner, created, creator, updated, updater) "
                + "VALUES (:id, :name, :description, :owner, :created, :creator, :updated, :updater)")
        .param("id", lrmItem.id())
        .param("name", lrmItem.name())
        .param("description", lrmItem.description())
        .param("owner", lrmItem.owner())
        .param("created", toJavaInstant(lrmItem.created()))
        .param("creator", lrmItem.creator())
        .param("updated", toJavaInstant(lrmItem.updated()))
        .param("updater", lrmItem.updater())
        .update();
    return lrmItem.id();
  }

  @Override
  public int update(LrmItem lrmItem) {
    return jdbcClient
        .sql(
            "UPDATE item SET name = :name, description = :description, updated = :updated WHERE id = :id")
        .param("name", lrmItem.name())
        .param("description", lrmItem.description())
        .param("updated", toJavaInstant(Clock.System.INSTANCE.now()))
        .param("id", lrmItem.id())
        .update();
  }

  @Override
  public int updateName(LrmItem lrmItem) {
    return jdbcClient
        .sql("UPDATE item SET name = :name, updated = :updated WHERE id = :id")
        .param("name", lrmItem.name())
        .param("updated", toJavaInstant(Clock.System.INSTANCE.now()))
        .param("id", lrmItem.id())
        .update();
  }

  @Override
  public int updateDescription(LrmItem lrmItem) {
    return jdbcClient
        .sql("UPDATE item SET description = :description, updated = :updated WHERE id = :id")
        .param("description", lrmItem.description())
        .param("updated", toJavaInstant(Clock.System.INSTANCE.now()))
        .param("id", lrmItem.id())
        .update();
  }

  private static List<LrmItem> mapRowsToLrmItems(List<ItemRow> rows) {
    Map<UUID, List<LrmListSuccinct>> listsByItem =
        rows.stream()
            .filter(row -> row.listId() != null)
            .collect(
                Collectors.groupingBy(
                    ItemRow::itemId,
                    LinkedHashMap::new,
                    Collectors.mapping(
                        row -> new LrmListSuccinct(row.listId(), row.listName()),
                        Collectors.toList())));

    Map<UUID, LrmItem> distinctByItemId = new LinkedHashMap<>();
    for (ItemRow row : rows) {
      distinctByItemId.putIfAbsent(row.itemId(), toLrmItem(row));
    }

    List<LrmItem> result = new ArrayList<>();
    for (LrmItem item : distinctByItemId.values()) {
      List<LrmListSuccinct> lists = listsByItem.get(item.id());
      Set<LrmListSuccinct> sortedLists =
          lists == null
              ? Set.of()
              : lists.stream()
                  .sorted(Comparator.comparing(LrmListSuccinct::name))
                  .collect(Collectors.toCollection(LinkedHashSet::new));
      result.add(item.withLists(sortedLists));
    }
    return result;
  }

  private static LrmItem toLrmItem(ItemRow row) {
    return new LrmItem(
        row.itemId(),
        row.itemName(),
        row.itemDescription(),
        row.itemOwner(),
        new kotlinx.datetime.Instant(row.itemCreated()),
        row.itemCreator(),
        new kotlinx.datetime.Instant(row.itemUpdated()),
        row.itemUpdater(),
        Set.of());
  }

  private static Instant toJavaInstant(kotlinx.datetime.Instant instant) {
    return ConvertersKt.toJavaInstant(instant);
  }

  private record ItemRow(
      UUID itemId,
      String itemName,
      String itemDescription,
      String itemOwner,
      Instant itemCreated,
      String itemCreator,
      Instant itemUpdated,
      String itemUpdater,
      UUID listId,
      String listName) {}
}

File: src/main/kotlin/net/flyingfishflash/loremlist/persistence/LrmItemRepositoryRdbms.kt
package net.flyingfishflash.loremlist.persistence

import kotlinx.datetime.Clock.System.now
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItem
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemRepository
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct
import org.jetbrains.exposed.sql.Query
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SqlExpressionBuilder.inList
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.andWhere
import org.jetbrains.exposed.sql.count
import org.jetbrains.exposed.sql.deleteAll
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.orWhere
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.statements.UpdateStatement
import org.jetbrains.exposed.sql.update
import org.springframework.stereotype.Repository
import java.util.UUID

@Repository
class LrmItemRepositoryRdbms : LrmItemRepository {
  private val repositoryTable = LrmItemTable

  override fun countByOwner(owner: String): Long {
    val idCount = repositoryTable.id.count()
    return repositoryTable.select(idCount).where { repositoryTable.owner eq owner }.first()[idCount]
  }

  override fun delete(): Int = repositoryTable.deleteAll()

  override fun deleteById(ids: Set<UUID>): Int = repositoryTable.deleteWhere { repositoryTable.id inList ids }

  override fun deleteByOwnerAndId(id: UUID, owner: String): Int = repositoryTable.deleteWhere {
    repositoryTable.id eq id and (repositoryTable.owner eq owner)
  }

  override fun findByOwner(owner: String): List<LrmItem> {
    return mapQueryResultToLrmItems(
      (repositoryTable leftJoin LrmListItemsTable leftJoin LrmListTable)
        .selectAll()
        .byOwner(owner),
    )
  }

  override fun findByOwnerAndIdOrNull(id: UUID, owner: String): LrmItem? {
    val lrmItems = mapQueryResultToLrmItems(
      (repositoryTable leftJoin LrmListItemsTable leftJoin LrmListTable)
        .selectAll()
        .byOwnerAndId(owner, id),
    ).distinct()
    check(lrmItems.size <= 1) { "This query should return no more than one record." }
    return lrmItems.firstOrNull()
  }

  override fun findByOwnerAndHavingNoListAssociations(owner: String): List<LrmItem> {
    val result = (repositoryTable leftJoin LrmListItemsTable)
      .selectAll()
      .where { repositoryTable.owner eq owner }
      .andWhere { LrmListItemsTable.item.isNull() }
      .map { it.toLrmItem() }
    return result
  }

  override fun findByOwnerAndHavingNoListAssociations(owner: String, listId: UUID): List<LrmItem> {
    val result = (repositoryTable leftJoin LrmListItemsTable)
      .selectAll()
      .where { repositoryTable.owner eq owner }
      .andWhere { LrmListItemsTable.list neq listId }.orWhere { LrmListItemsTable.list.isNull() }
      .map { it.toLrmItem() }
    return result
  }

  override fun findIdsByOwnerAndIds(itemIdCollection: List<UUID>, owner: String): List<UUID> {
    return repositoryTable.select(repositoryTable.id)
      .where { repositoryTable.id inList itemIdCollection }
      .andWhere { repositoryTable.owner eq owner }
      .map { it[repositoryTable.id] }
      .toList()
  }

  override fun notFoundByOwnerAndId(itemIdCollection: List<UUID>, owner: String): Set<UUID> {
    val resultIdCollection = findIdsByOwnerAndIds(itemIdCollection = itemIdCollection, owner = owner)
    return (itemIdCollection subtract resultIdCollection.toSet())
  }

  override fun insert(lrmItem: LrmItem): UUID {
    repositoryTable.insert {
      it[id] = lrmItem.id
      it[name] = lrmItem.name
      it[description] = lrmItem.description
//      it[quantity] = lrmItem.quantity
      it[owner] = lrmItem.owner
      it[created] = lrmItem.created
      it[creator] = lrmItem.creator
      it[updated] = lrmItem.updated
      it[updater] = lrmItem.updater
    }
    return lrmItem.id
  }

  override fun update(lrmItem: LrmItem): Int = updateItem(lrmItem) {
    it[repositoryTable.name] = lrmItem.name
    it[repositoryTable.description] = lrmItem.description
//    it[repositoryTable.quantity] = lrmItem.quantity
  }

  override fun updateName(lrmItem: LrmItem): Int = updateItem(lrmItem) { it[repositoryTable.name] = lrmItem.name }

  override fun updateDescription(lrmItem: LrmItem): Int = updateItem(lrmItem) { it[repositoryTable.description] = lrmItem.description }

  private fun updateItem(lrmItem: LrmItem, updateAction: (UpdateStatement) -> Unit): Int {
    return repositoryTable.update({ repositoryTable.id eq lrmItem.id }) {
      updateAction(it)
      it[repositoryTable.updated] = now()
    }
  }

  private fun Query.byOwner(owner: String): Query {
    return this.where { repositoryTable.owner eq owner }
  }

  private fun Query.byOwnerAndId(owner: String, id: UUID): Query {
    return this.byOwner(owner).andWhere { repositoryTable.id eq id }
  }

  private fun mapQueryResultToLrmItems(query: Query): List<LrmItem> {
    val resultRows = query.toList()

    @Suppress("SENSELESS_COMPARISON")
    val listsByItems = resultRows
      .filter { it[LrmListTable.id] != null }
      .groupBy(
        keySelector = { it[repositoryTable.id] },
        valueTransform = {
          LrmListSuccinct(
            id = it[LrmListTable.id],
            name = it[LrmListTable.name],
          )
        },
      )

    return resultRows
      .distinctBy { it[repositoryTable.id] }
      .map {
        it.toLrmItem().copy(
          lists = listsByItems[it[repositoryTable.id]]?.sortedBy { list -> list.name }?.toSet() ?: emptySet(),
        )
      }
  }

  fun ResultRow.toLrmItem(): LrmItem {
    return LrmItem(
      id = this[repositoryTable.id],
      name = this[repositoryTable.name],
      description = this[repositoryTable.description],
      owner = this[repositoryTable.owner],
      created = this[repositoryTable.created],
      creator = this[repositoryTable.creator],
      updated = this[repositoryTable.updated],
      updater = this[repositoryTable.updater],
    )
  }
}
